tests/test_command_shell.py

```diff
@@ -113,3 +114,13 @@ def test_dns_failures(self):
                                        check_code=False)
         self.assertEqual(errcode, 1, out or err)
         self.assertIn(b'DNS lookup failed', err)
+
+    @defer.inlineCallbacks
+    def test_dns_idna(self):
+        """Twisted 17.1+ passes Unicode domain to getHostByName() in Python 2.7,
+        and Scrapy's resolver must be able to handle them."""
+        # this is the same as http://δπθ.localhost
+        with mock.patch('socket.gethostbyname', side_effect='127.0.0.1'):
```

Comment:
Question: Should you be mocking `socket` in `twisted.internet.base.socket` [here](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/base.py#L11)? I did a brief search on how scrapy resolves dns and it seem to me like it calls `ThreadedResolver` over [here](https://github.com/scrapy/scrapy/blob/245287fd5a357655f7d7e7f2e333e2e7c5794262/scrapy/resolver.py#L10).  `socket.hostname` seems to be called [here](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/base.py#L282).

I'm just taking a random guess. I haven't tested it nor have I validated my source?